code/datums/weather/weather.dm

```diff
@@ -105,28 +146,68 @@
 /datum/weather/proc/telegraph()
 	if(stage == STARTUP_STAGE)
 		return
-	SEND_GLOBAL_SIGNAL(COMSIG_WEATHER_TELEGRAPH(type), src)
 	stage = STARTUP_STAGE
+	setup_weather_areas(impacted_areas)
+
+	if(weather_flags & (WEATHER_TURFS|WEATHER_THUNDER))
+		setup_weather_turfs()
+
+	SEND_GLOBAL_SIGNAL(COMSIG_WEATHER_TELEGRAPH(type), src)
+
+	weather_duration = rand(weather_duration_lower, weather_duration_upper)
+	SSweather.processing |= src
+	update_areas()
+	if(telegraph_duration)
+		send_alert(telegraph_message, telegraph_sound, telegraph_sound_vol)
+	addtimer(CALLBACK(src, PROC_REF(start)), telegraph_duration)
+
+/datum/weather/proc/setup_weather_areas(list/selected_areas)
+	if(length(selected_areas))
+		return // impacted areas already been setup
+
 	var/list/affectareas = list()
 	for(var/area/selected_area as anything in get_areas(area_type))
 		affectareas += selected_area
 	for(var/area/protected_area as anything in protected_areas)
 		affectareas -= get_areas(protected_area)
 	for(var/area/affected_area as anything in affectareas)
-		if(protect_indoors && !affected_area.outdoors)
+		if(!(weather_flags & WEATHER_INDOORS) && !affected_area.outdoors)
 			continue
 
 		for(var/z in impacted_z_levels)
 			if(length(affected_area.turfs_by_zlevel) >= z && length(affected_area.turfs_by_zlevel[z]))
-				impacted_areas |= affected_area
+				selected_areas |= affected_area
 				continue
 
-	weather_duration = rand(weather_duration_lower, weather_duration_upper)
-	SSweather.processing |= src
-	update_areas()
-	if(telegraph_duration)
-		send_alert(telegraph_message, telegraph_sound, telegraph_sound_vol)
-	addtimer(CALLBACK(src, PROC_REF(start)), telegraph_duration)
+/datum/weather/proc/setup_weather_turfs()
+	for(var/area/weather_area as anything in impacted_areas)
+		for(var/z in impacted_z_levels)
+			for(var/turf/valid_weather_turf as anything in weather_area.get_turfs_by_zlevel(z))
+				// applying weather effects to solid walls is a waste since nothing will happen
+				if(isclosedturf(valid_weather_turf))
+					continue
+				// same logic for space and openspace turfs which should boost performance a ton
+				// note - mobs in space/openspace turfs still have weather affects applied to them if they are in a affected area
+				if(is_space_or_openspace(valid_weather_turf))
+					continue
+				// solid windows are also worth skipping
+				var/obj/structure/window/window = locate() in valid_weather_turf
+				if(window?.fulltile)
+					continue
+
+				weather_turfs += valid_weather_turf
+
+	var/total_turfs = length(weather_turfs)
+
+	if(!total_turfs || !(weather_flags & (WEATHER_TURFS|WEATHER_THUNDER)))
+		return
```

Comment:
why do we return if it's 0 doesn't that mean the var SHOULD be 0?

Reply:
Both `weather_turfs_per_tick` and `thunder_turfs_per_tick` default to 0 so they don't need to be set if there are no valid turfs in the weather areas. (which can happen if it's all walled off or space)

Reply:
oook